luaui/Widgets/gui_comselect.lua

```diff
@@ -0,0 +1,128 @@
+function widget:GetInfo()
+	return {
+		name = "Commander Selector",
+		desc = "Allows for proper keybind selection of any flavour of commanders",
+		author = "Hornet, with some ZK raiding",
+		date = "03 June 2024",
+		license = "GNU GPL, v2 or later",
+		layer = 1001,
+		enabled = true,
+	}
+end
+
+
+
+local myTeamID = Spring.GetMyTeamID()
+local spGetUnitTeam = Spring.GetUnitTeam
+local spGetMyTeamID = Spring.GetMyTeamID
+local spSendCommands = Spring.SendCommands
+
+local specOld = false
+local spec = false
+local team
+
+
+local commanderList = {}
+local commanderIds = {}
+for udid, ud in pairs(UnitDefs) do
+	if ud.customParams and ud.customParams.iscommander then
+		commanderIds[udid] = udid
+	end
+end
+
+function widget:MetaUnitAdded(unitID, unitDefID, unitTeam)
+	if (not myTeamID or unitTeam ~= myTeamID) then
+		return
+	end
+	if commanderIds[unitDefID] then
+		table.insert(commanderList, unitID)
+		--Spring.Echo('hornet comsel adding com')
+	end
+end
+
+
+function widget:MetaUnitRemoved(unitID, unitDefID, unitTeam)
+	if commanderList[unitID] then
+		commanderList[unitID] = nil
+	end
+end
+
+
+
+
+
+-- comm selection functionality
+local commIndex = 1
+local function SelectComm()
+	--Spring.Echo('hornet comsel tab')
+	local commCount = #commanderList
+
+	if commCount <= 0 then
+		--Spring.Echo('hornet comsel 0 coms')
+		return
+	end
+	
+	-- This check deals with the case of spectators selecting
+	-- teams with different numbers of commanders.
+	if commCount < commIndex then
+		commIndex = commCount
+	end
+	
+	local unitID
+	-- Loop long enough to check every commander.
+	-- The most recently Ctrl+C selected commander is checked last.
+	-- Select the first non-selected commander encountered.
+	for i = 1, commCount do
+		unitID = commanderList[commIndex]
+		commIndex = commIndex + 1
+		if commIndex > commCount then
+			commIndex = 1
+		end
+		if not Spring.IsUnitSelected(unitID) then
+			break
+		end
+	end
+	
+	local alt, ctrl, meta, shift = Spring.GetModKeyState()
+	Spring.SelectUnitArray({unitID}, shift)
+	if not shift then
+		local x, y, z = Spring.GetUnitPosition(unitID)
+		Spring.SetCameraTarget(x, y, z)
+	end
+end
+
+
+function widget:PlayerChanged()
+	local spectatorFollowing = Spring.GetMyTeamID()
+
+	if spectatorFollowing ~= nil then
+		commanderList = {}
+
+		local units = Spring.GetAllUnits()
+		for i = 1, #units do
+			local unitID = units[i]
+			local unitOwner = Spring.GetUnitTeam(unitID)
+
+			if unitOwner == spectatorFollowing then
+				local unitDefID = Spring.GetUnitDefID(unitID)
+
+				if commanderIds[unitDefID] then
+					table.insert(commanderList, unitID)
+				end
+
+			end
+		end
+	end
+
+  end
```

Comment:
`local spectatorFollowing = Spring.GetMyTeamID()` is pretty unfortunate because it applies to playing players as well, not just specs. There's also no point getting all teams' units. I would do it like this (which also handles `/luaui reload`):

```suggestion
local function RefreshList()
	local spGetUnitDefID = Spring.GetUnitDefID

	commanderList = {}
	local count = 0
	local units = Spring.GetTeamUnits(myTeamID)
	for i = 1, #units do
		local unitID = units[i]
		if commanderIds[spGetUnitDefID(unitID)] then
			count = count + 1
			commanderList[count] = unitID
		end
	end
end

function widget:PlayerChanged()
	local teamID = Spring.GetMyTeamID()

	if teamID ~= myTeamID then
		myTeamID = teamID
		RefreshList()
	end
end

function widget:Initialize()
	myTeamID = Spring.GetMyTeamID()
	RefreshList()
end
```

Reply:
I wasn't aware a player could trigger playerchanged() though, surely only spectators can?

Reply:
`PlayerChanged` runs in these cases:
* you, as a spectator, change which team you are spectating.
* anybody, including but not limited to yourself, changes from player to spectator (i.e. resigns). This can happen even if you are a player.
* a player changes their team, or a spectator becomes a player. AFAIK these aren't currently applicable to BAR.

Reply:
Noted then, will roll your code in and do some testing, thanks muchly